src/org/mozilla/javascript/JavaMembers.java

```diff
@@ -218,80 +108,192 @@ static String liveConnectSignature(Class<?>[] argTypes)
+            } else {
+                ObjArray overloadedMethods;
+                if (value instanceof ObjArray) {
+                    overloadedMethods = (ObjArray)value;
+                } else {
+                    if (!(value instanceof Method)) Kit.codeBug();
+                    // value should be instance of Method as at this stage
+                    // staticMembers and members can only contain methods
+                    overloadedMethods = new ObjArray();
+                    overloadedMethods.add(value);
+                    ht.put(name, overloadedMethods);
+                }
+                overloadedMethods.add(method);
+            }
+        }
+        }
 
-        if (methodOrCtor != null) {
-            Scriptable prototype =
-                ScriptableObject.getFunctionPrototype(scope);
+        Field getStaticField(final String key) {
+            return staticFieldMap.get(key);
+        }
 
-            if (methodOrCtor.isCtor()) {
-                NativeJavaConstructor fun =
-                    new NativeJavaConstructor(methodOrCtor);
-                fun.setPrototype(prototype);
-                member = fun;
-                ht.put(name, fun);
-            } else {
-                String trueName = methodOrCtor.getName();
-                member = ht.get(trueName);
+        Field getInstField(final String key) {
+            return instFieldMap.get(key);
+        }
 
-                if (member instanceof NativeJavaMethod &&
-                    ((NativeJavaMethod)member).methods.length > 1 ) {
-                    NativeJavaMethod fun =
-                        new NativeJavaMethod(methodOrCtor, name);
-                    fun.setPrototype(prototype);
-                    ht.put(name, fun);
-                    member = fun;
+        private static final <T> Set<T> intersection(Collection<T> a, Collection<T> b){
+            final Set<T> ret = new HashSet<T>();
+            for(T v : a) {
+                if(b.contains(v)) {
+                    ret.add(v);
                 }
             }
+            return ret;
         }
+    }
 
-        return member;
+    private static final Map<String,Object> cache = new ConcurrentHashMap<String, Object>();
+    private static final Map<Class,ClassReflectBean>[] cacheArray = new Map[]{
+        new ConcurrentHashMap<Class,ClassReflectBean>(),
+        new ConcurrentHashMap<Class,ClassReflectBean>(),
+        new ConcurrentHashMap<Class,ClassReflectBean>()
+    };
+
+    private static Map<Class,ClassReflectBean> getCache(final Class<?> clazz, final boolean includeProtected, final boolean includePrivate){
+        int i=0;
+        if(includeProtected) {
+            i++;
+        }
+        if(includePrivate) {
+            i++;
+        }
+        return cacheArray[i];
+    }
+
+    //rhino_JavaMembers_reflect_cache_on=false for disable
+    private static final boolean CACHE_ON = !"false".equals(getProperty("rhino_JavaMembers_reflect_cache_on","true"));
```

Comment:
I would rather do this kind of optional stuff by following the existing pattern, which uses constants in the Context class and code in ContextFactory to select this feature. But I am glad that you have thought of making it optional!

Reply:
In  my own project I perfer SPI(java.util.ServiceLoader) way for the configuration

> I would rather do this kind of optional stuff by following the existing pattern, which uses constants in the Context class and code in ContextFactory to select this feature. But I am glad that you have thought of making it optional!

In  my own project I perfer SPI(java.util.ServiceLoader) way for the configuration, so that the app can do the configuration it's own way:)